conda/common/configuration.py

```diff
@@ -1049,51 +1060,130 @@ def load(self, name, match):
             validation=self._validation)
 
 
+#: These are the allowable types for a SequenceParameter
+SequenceElementTypes = Union[MapParameter, PrimitiveParameter]
+
+
 class SequenceParameter(Parameter):
     """
-    Parameter type for a Configuration class that holds a sequence (i.e. list) of Parameters.
+    The SequenceParameter class allows you to define a flexible sequence that
+    can accept multiple data types. The simplest use case is accepting a single
+    PrimitiveParameter, but you can also pass in a map_type that will allow for
+    defining a MapParameter that is also permissible in the sequence.
+
+    More examples of what this looks like are available in ``tests/common/test_configuration.py``
     """
-    _type = tuple
 
-    def __init__(self, element_type, default=(), validation=None, string_delimiter=','):
+    def __init__(
+        self,
+        default: Sequence | None = None,
+        primitive_type: PrimitiveParameter = None,
+        map_type: SequenceElementTypes = None,
```

Comment:
I think we should make this even more abstract. This is currently too specific to what we are trying to accomplish with the `channels` context argument. Something like:

```
SequenceElementTypes = PrimitiveParameter | MapParameter | SequenceParameter

class SequenceParameter(Parameter):
    def __init__(
        self,
        element_type: SequenceElementTypes | list[SequenceElementTypes],
        default: Sequence | None = None,
        validation=None,
        string_delimiter: str = ",",
    ):
        ...
```

Reply:
Hi @kenodegard,

I spent a good amount of time deliberating your suggestion and going back over a lot of the code contained within `conda/common/configuration.py` and `conda/base/context.py`.

What I concluded is that although it's alluring to overload the `element_type` parameter, this parameter really only wants to point at a single type. Let me explain what I mean with an example...

Suppose we have the following configuration defined:

```python
from conda.common.configuration import (
    Configuration,
    SequenceParameter,
    PrimitiveParameter,
   MapParameter,
    ParameterLoader, 

class MyConfig(Configuration):
    complex_seq = ParameterLoader(
        SequenceParameter(element_type=(
            PrimitiveParameter(0, element_type=int),
            PrimitiveParameter("", element_type=str),
            MapParameter(PrimitiveParameter("", element_type=str))
        ))
    )
```

We've added in `PrimitiveParameter` twice, which may cause undefined behavior. When we run across a number, do we parse it as a string or an integer? Defining `element_type` as a sequence of `Parameter` types leaves us open to this type of ambiguity in our API.

But, what if we just want to enforce that users only pass in one of each type. In this case, we would have to implement some ugly validation logic within these objects to verify that. There must be a better way...

This is why I opted to create two new keyword arguments for this `Parameter` type. To avoid ambiguities like in the above example, we want to establish a one-to-one relation between what is found in the configuration and how it ultimately ends up getting parsed.

I would gladly add an additional `sequence_type` so that it supports all possible values, but I figured that would be a little overkill for this pull request. I hope that I have documented the class well enough that a would-be future developer wouldn't have a difficult time adding this behavior in the future if it is ever needed.

Let me know what you think and if I have adequately convinced you 😎 .  I'm always happy to discuss this code and the rationale behind the changes more.

Reply:
I've always preferred the (sometimes extreme) abstractions for stuff like this since I've always benefitted from it eventually. 🤷🏼‍♂️

I would rely on the order of `element_types` to dictate the parsing precedence, so parsing `"0"` will always return a `0`. If the `PrimitiveParameter` were switched, we'd expect a `"0"` to always be `"0"`.

This is so close to implementing true homogeneous and heterogenous sequence types that not implementing it IMO results in a more complicated solution that will be harder to document.

But I don't care to die on this hill. I'll let others chime in and you can make a decision from there.

Reply:
@kenodegard ,

I went back and made some revisions. I have more-or-less implemented the type of API you would like for `SequenceParameter` and now we can have a combination of primitive values, maps and sequences.

I opted to use a `element_types` parameter that is just a sequence rather than the sequence or single value function signature you had suggested with the `element_type` parameter.

It still works the same "under the hood", but I've made sure to make class attributes like `_map_type`, `_primitive_type` and `_sequence_type` protected so people know they are not supposed to use these directly.

Thanks for your help in discussing/evolving this pull request with me! Let me know if there's still anything else that needs to be resolved, or whether the newest batch of changes aligns with what you had imagined.

Reply:
@kenodegard ,

You can disregard part of that previous comment. After seeing the code diff once more after my changes, I realized it would be better to have that `element_type` argument and give it the ability to accept either a single Parameter type or a sequence of them.

This should hopefully be everything you are looking for now.

Cheers ✌️